scripts/baseboxd-knet-reset.py

```
@@ -0,0 +1,92 @@
+# file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+
+from ctypes import *
+import fcntl
+import os
+
+class bkn_ioctl_t(Structure):
+    _fields_ = [("rc", c_int),
+                ("len", c_int),
+                ("bufsz", c_int),
+                ("reserved", c_int),
+                ("buf", c_uint64)]
+
+class kcom_msg_hdr_t(Structure):
+    _fields_ = [("type", c_uint8),
+                ("opcode", c_uint8),
+                ("seqno", c_uint8),
+                ("status", c_uint8),
+                ("unit", c_uint8),
+                ("reserved", c_uint8),
+                ("id", c_uint16)]
+
+class kcom_msg_list_t(Structure):
+    _fields_ = [("hdr", kcom_msg_hdr_t),
+                ("fcnt", c_uint16),
+                ("id", c_uint16 * 128)]
+
+KCOM_MSG_TYPE_CMD = 1
+
+KCOM_M_NETIF_DESTROY = 12
+KCOM_M_NETIF_LIST = 13
+
+KCOM_M_FILTER_DESTROY = 22
+KCOM_M_FILTER_LIST = 23
+
+# Used for NETIF_DESTROY and FILTER_DESTROY.
+#
+# Both use only the id field of the header to identify the filter/netif
+# to destroy.
+def delete_kcom_object(fd, obj_id, opcode):
+    hdr = kcom_msg_hdr_t()
+    hdr.type = KCOM_MSG_TYPE_CMD
+    hdr.opcode = opcode
+    hdr.unit = 0
+    hdr.id = obj_id
+
+    io = bkn_ioctl_t()
+    io.len = sizeof(hdr)
+    io.buf = addressof(hdr)
+
+    fcntl.ioctl(fd, 0, io)
+
+# Used for NETIF_LIST and FILTER_LIST
+#
+# Technically they use different structs passed to the kernel module,
+# practically they have identical layouts, so we can just use one type.
+def get_obj_list(fd, opcode):
+    obj_list = kcom_msg_list_t()
+    obj_list.hdr.type = KCOM_MSG_TYPE_CMD
+    obj_list.hdr.opcode = opcode
+    obj_list.hdr.unit = 0
+
+    io = bkn_ioctl_t()
+    io.len = sizeof(obj_list)
+    io.buf = addressof(obj_list)
+
+    fcntl.ioctl(fd, 0, io)
+
+    return obj_list
+
+if os.path.exists("/dev/linux-bcm-knet"):
+    fd = os.open("/dev/linux-bcm-knet", os.O_RDWR | os.O_NONBLOCK)
```

```suggestion
    with  os.open("/dev/linux-bcm-knet", os.O_RDWR | os.O_NONBLOCK) as fd:
       ....
```

---

Actually, if you want to go that way, drop "if os.path.exists("/dev/linux-bcm-knet")". It's not pythonic.

---

So what's the alternative?

---

Then we would need to do a check in the exception handler for it to know whether this is an expected exception or not (if we are running offswitch, then there will be no /dev/linux-bcm-knet, so it failing to open is expected and shouldn't be presented as an error and silently ignored).

Not sure if this is better.

---

```python
try:
    fd = os.open("/dev/linux-bcm-knet", os.O_RDWR | os.O_NONBLOCK)
except FileNotFoundError:
    pass
```
I'm not sure it's _better_, but it is closer to what Python code is supposed to work like.

---

After thinking about this, I think this one is nice since I can then add a comment before the pass that this is fine and isn't an error (which I just did).